File: ios/ReservationApp/ReservationApp/Presentation/ViewModels/ProfileViewModel.swift
//
//  ProfileViewModel.swift
//  ReservationApp
//
//  Created by Mehmet Ali Kısacık on 13.11.2025.
//

import Foundation
import SwiftUI
import Combine

@MainActor
class ProfileViewModel: ObservableObject {
    @Published var user: User?
    @Published var isLoading = false
    @Published var isSaving = false
    @Published var errorMessage: String?
    @Published var successMessage: String?
    
    // Form data
    @Published var name: String = ""
    @Published var email: String = ""
    @Published var department: String = ""
    
    private let authRepository: AuthRepositoryProtocol
    private let keychainManager: KeychainManager
    private let authStateManager: AuthStateManager
    
    init(
        authRepository: AuthRepositoryProtocol = AuthRepository(),
        keychainManager: KeychainManager = KeychainManager.shared,
        authStateManager: AuthStateManager = AuthStateManager.shared
    ) {
        self.authRepository = authRepository
        self.keychainManager = keychainManager
        self.authStateManager = authStateManager
    }
    
    func loadProfile() async {
        isLoading = true
        errorMessage = nil
        
        do {
            let userData = try await authRepository.getCurrentUser()
            self.user = userData
            self.name = userData.name
            self.email = userData.email
            self.department = userData.department ?? ""
        } catch {
            errorMessage = "Profil bilgileri yüklenirken bir hata oluştu."
        }
        
        isLoading = false
    }
    
    func saveProfile() async {
        guard !name.trimmingCharacters(in: .whitespaces).isEmpty else {
            errorMessage = "Ad alanı zorunludur."
            return
        }
        
        isSaving = true
        errorMessage = nil
        successMessage = nil
        
        do {
            let request = UpdateUserProfileRequest(name: name, department: department.isEmpty ? nil : department)
            let updatedUser = try await authRepository.updateUserProfile(request: request)
            self.user = updatedUser
            self.name = updatedUser.name
            self.email = updatedUser.email
            self.department = updatedUser.department ?? ""
            successMessage = "Profil başarıyla güncellendi"
        } catch {
            errorMessage = "Profil kaydedilirken bir hata oluştu."
        }
        
        isSaving = false
    }
    
    func logout() {
        // Delete token from keychain (following frontend pattern - no API call needed)
        keychainManager.deleteToken()
        
        // Clear user data
        self.user = nil
        self.name = ""
        self.email = ""
        self.department = ""
        
        // Update authentication state - this will trigger RootView to show OnboardingView
        authStateManager.setAuthenticated(false)
    }
}

File: android/feature/profile/src/main/java/com/reservationapp/feature/profile/viewmodel/ProfileViewModel.kt
package com.reservationapp.feature.profile.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.reservationapp.core.common.AuthStateManager
import com.reservationapp.core.common.Result
import com.reservationapp.domain.model.UpdateUserProfileRequest
import com.reservationapp.domain.model.User
import com.reservationapp.domain.repository.AuthRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.launch

class ProfileViewModel(
    private val authRepository: AuthRepository,
    private val authStateManager: AuthStateManager
) : ViewModel() {

    val currentUser: StateFlow<User?> = authStateManager.currentUser.map { it as? User }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5000),
        initialValue = null
    )

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _successMessage = MutableStateFlow<String?>(null)
    val successMessage: StateFlow<String?> = _successMessage.asStateFlow()

    // Editable fields
    val name = MutableStateFlow("")
    val email = MutableStateFlow("")
    val department = MutableStateFlow<String?>(null)

    init {
        loadUserProfile()
    }

    private fun loadUserProfile() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            when (val result = authRepository.getCurrentUser()) {
                is Result.Success -> {
                    val user = result.data
                    name.value = user.name
                    email.value = user.email
                    department.value = user.department
                    authStateManager.setCurrentUser(user)
                }
                is Result.Error -> {
                    _errorMessage.value = "Profil bilgileri yüklenemedi"
                }
                is Result.Loading -> {
                    // Handle loading
                }
            }

            _isLoading.value = false
        }
    }

    fun updateProfile() {
        if (name.value.isEmpty() || email.value.isEmpty()) {
            _errorMessage.value = "İsim ve email alanları zorunludur"
            return
        }

        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            _successMessage.value = null

            val request = UpdateUserProfileRequest(
                name = name.value,
                email = email.value,
                department = department.value
            )

            when (val result = authRepository.updateProfile(request)) {
                is Result.Success -> {
                    val updatedUser = result.data
                    authStateManager.setCurrentUser(updatedUser)
                    _successMessage.value = "Profil başarıyla güncellendi"
                }
                is Result.Error -> {
                    _errorMessage.value = "Profil güncellenirken bir hata oluştu"
                }
                is Result.Loading -> {
                    // Handle loading
                }
            }

            _isLoading.value = false
        }
    }

    fun refresh() {
        loadUserProfile()
    }

    fun resetForm() {
        loadUserProfile()
    }

    fun logout() {
        viewModelScope.launch {
            // Clear secure storage
            authRepository.logout()
            
            // Clear user data from form
            name.value = ""
            email.value = ""
            department.value = null
            
            // Clear auth state - this will trigger navigation to login
            authStateManager.clearUser()
        }
    }

    fun clearMessages() {
        _errorMessage.value = null
        _successMessage.value = null
    }
}

